src/emulate.c

```diff
@@ -593,18 +613,314 @@ static void block_translate(riscv_t *rv, block_t *block)
         ir->tailcall = next_ir->tailcall;                                  \
     }
 
+static bool parse_memset(riscv_t *rv)
+{
+    static const uint32_t memset_insn[] = {
+        0x00f00313, /* li	t1,15 */
+        0x00050713, /* mv	a4,a0 */
+        0x02c37e63, /* bgeu	t1,a2,0x11828 */
+        0x00f77793, /* and	a5,a4,15 */
+        0x0a079063, /* bnez	a5,0x11894 */
+        0x08059263, /* bnez	a1,0x1187c */
+        0xff067693, /* and	a3,a2,-16 */
+        0x00f67613, /* and	a2,a2,15 */
+        0x00e686b3, /* add	a3,a3,a4 */
+        0x00b72023, /* sw	a1,0(a4) */
+        0x00b72223, /* sw	a1,4(a4) */
+        0x00b72423, /* sw	a1,8(a4) */
+        0x00b72623, /* sw	a1,12(a4) */
+        0x01070713, /* add	a4,a4,16 */
+        0xfed766e3, /* bltu	a4,a3,0x11808 */
+        0x00061463, /* bnez	a2,0x11828 */
+        0x00008067, /* ret */
+        0x40c306b3, /* sub	a3,t1,a2 */
+        0x00269693, /* sll	a3,a3,0x2 */
+        0x00000297, /* auipc	t0,0x0 */
+        0x005686b3, /* add	a3,a3,t0 */
+        0x00c68067, /* jr	12(a3) */
+        0x00b70723, /* sb	a1,14(a4) */
+        0x00b706a3, /* sb	a1,13(a4) */
+        0x00b70623, /* sb	a1,12(a4) */
+        0x00b705a3, /* sb	a1,11(a4) */
+        0x00b70523, /* sb	a1,10(a4) */
+        0x00b704a3, /* sb	a1,9(a4) */
+        0x00b70423, /* sb	a1,8(a4) */
+        0x00b703a3, /* sb	a1,7(a4) */
+        0x00b70323, /* sb	a1,6(a4) */
+        0x00b702a3, /* sb	a1,5(a4) */
+        0x00b70223, /* sb	a1,4(a4) */
+        0x00b701a3, /* sb	a1,3(a4) */
+        0x00b70123, /* sb	a1,2(a4) */
+        0x00b700a3, /* sb	a1,1(a4) */
+        0x00b70023, /* sb	a1,0(a4) */
+        0x00008067, /* ret */
+        0x0ff5f593, /* zext.b	a1,a1 */
+        0x00859693, /* sll	a3,a1,0x8 */
+        0x00d5e5b3, /* or	a1,a1,a3 */
+        0x01059693, /* sll	a3,a1,0x10 */
+        0x00d5e5b3, /* or	a1,a1,a3 */
+        0xf6dff06f, /* j	0x117fc */
+        0x00279693, /* sll	a3,a5,0x2 */
+        0x00000297, /* auipc	t0,0x0 */
+        0x005686b3, /* add	a3,a3,t0 */
+        0x00008293, /* mv	t0,ra */
+        0xfa0680e7, /* jalr	-96(a3) */
+        0x00028093, /* mv	ra,t0 */
+        0xff078793, /* add	a5,a5,-16 */
+        0x40f70733, /* sub	a4,a4,a5 */
+        0x00f60633, /* add	a2,a2,a5 */
+        0xf6c378e3, /* bgeu	t1,a2,0x11828 */
+        0xf3dff06f, /* j	0x117f8 */
+    };
+    uint32_t tmp_pc = rv->PC;
+    for (int i = 0; i < 55; i++) {
+        const uint32_t insn = rv->io.mem_ifetch(tmp_pc);
+        if (insn != memset_insn[i])
+            return false;
+        tmp_pc += 4;
+    }
+    return true;
+}
+
+static bool parse_memset2(riscv_t *rv)
```

Comment:
Should be `detect_memset2`. Of course, it would be better if one routine can manipulate several patterns.